validator/sawtooth_validator/consensus/proxy.py

```diff
@@ -29,6 +33,73 @@ class UnknownBlock(Exception):
     """The given block could not be found."""
 
 
+class ConsensusActivationObserver(ChainObserver):
+    """Observes chain updates and activates new engines, when the settings
+    have been updated in state, as of the commit."""
+
+    def __init__(self, consensus_registry, consensus_notifier,
+                 settings_view_factory):
+        self._registry = consensus_registry
+        self._notifier = consensus_notifier
+        self._settings_view_factory = settings_view_factory
+
+    def chain_update(self, block, receipts):
+        # Check to see if there is an active engine at all:
+        if not self._registry.has_active_engine():
+            self._update_active_engine(block)
+            return
+
+        # Otherwise, check to see if the settings changed, before updating
+        # the active engine.
+        receipt_events = ReceiptEventExtractor(receipts).extract([
+            EventSubscription(event_type="setting/update")])
+
+        for event in receipt_events:
+            if self._is_consensus_algorithm_setting(event):
+                self._update_active_engine(block)
+                break
+
+    def _update_active_engine(self, block):
+        header = BlockHeader()
+        header.ParseFromString(block.header)
+        settings_view = self._settings_view_factory.create_settings_view(
+            header.state_root_hash)
+
+        conf_name = settings_view.get_setting(
+            'sawtooth.consensus.algorithm.name')
+        conf_version = settings_view.get_setting(
+            'sawtooth.consensus.algorithm.version')
+
+        if not self._registry.is_active_engine_name_version(
+                conf_name, conf_version):
+
+            # Deactivate the old engine, if necessary
+            old_engine_info = None
```

Comment:
This bit could be simpler by just notifying the deactivation first then activating and notifying the activation. But up to you.

Reply:
Notifying of deactivation cannot come before changing the active engine in the registry, because sending the notification doesn't alter the registry.  So I need to get the old engine, change the active engine, and then send the messages.